### ISSUE TITLE
undefined method `full_name' for s(:lvar, :m):#<Class:0x00>

### ISSUE THREAD

**Issue by backus** [troessner/reek#832]
Code to reproduce

```ruby
m = Module.new

module m::N
end
```

Commit which I think introduced the regression: 8bf26f9dea4ed16e3282bbf9d4ed1dc98af21133

**Comment by troessner** [comment]
Thanks for the report(s)! Oh boy, we really need to get #757 going ;)

**Comment by mvz** [comment]
Oh god why is that even allowed? Even worse: The following works and prints `Foo!`:

```ruby
module Bar; end

def blub(arg)
  arg == 3
end

module if blub(3)
    Bar
  else
    Qux
  end::Baz
  def foo
    puts 'Foo!'
  end
end

class Quuz
  include Bar::Baz
end

Quuz.new.foo
```

**Comment by backus** [comment]
> Oh god why is that even allowed?

lol

> The following works and prints `Foo!`:

shouldn't it?

**Comment by mvz** [comment]
> shouldn't it?

Well, if you want to namespace your modules with a 5 line if statement, it should :smile:.

Oddly, method definitions are much more limited: Anything more complex than a variable or argumentless method call is not accepted as a receiver.

**Comment by troessner** [comment]
>> module if blub(3)
    Bar
  else
    Qux
  end::Baz
  def foo
    puts 'Foo!'
  end
end

vs. 

>> shouldn't it?

No. No, it should not.

But I guess this discussing is moot, since Ruby won't care for it anyway :laughing:

**Comment by backus** [comment]
> > > shouldn't it?
> No. No, it should not.

Oh yeah I mean it is terrible that the language lets you do something like this. I was just double checking to make sure it at least made sense in terms of how the code evaluated.

**Comment by backus** [comment]
> Oh boy, we really need to get #757 going ;)

Yes I completely agree. In fact, I found this by running master against rubyspec.

I was playing with adding a corpus test last night based on what @mbj has configured for mutant. Please don't depend on me for the PR of course, I'm not sure if I have time, but for someone familiar with the reek codebase it really shouldn't be much work to port what mutant uses.

**Comment by troessner** [comment]
Fixed with the release of 3.9.0 just now.
